### PR TITLE
feat: Product Usage UI

### PR DESCRIPTION
**Pull request by ivan-degtiarenko**
## Description

Add UI for new product usage endpoints.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed
Screenshot 2023-08-15 at 16 12 52
Screenshot 2023-08-15 at 16 12 40
Screenshot 2023-08-15 at 16 12 29
Screenshot 2023-08-15 at 16 12 16
Screenshot 2023-08-15 at 16 12 02

### CODE REVIEW

**Comment by 0x656b694d** [comment]
* I don't see a "query" button. With the current storage implementation, the query for maximum usage may take some time: by default, the request returns maximum since ever;
* I don't see a "Today" button for the end of the time range. That could be helpful;
* Is it possible to erase time range after having it set?

**Comment by ivan-degtiarenko** [comment]
@0x656b694d 
> I don't see a "query" button. With the current storage implementation, the query for maximum usage may take some time: by default, the request returns maximum since ever;

I don't think "query" button is UX-friendly considering the query only consists of two parameters that you select by hand. Currently I set default query to last month - I think it's alright if the request takes some time to execute as it is not sent by default when person opens "System Health" - rather when they specifically press "Show product usage"

> I don't see a "Today" button for the end of the time range. That could be helpful

Unfortunately, we can't embed such button within calendar itself - so it would have to stand outside. Considering we don't forbid user to request usage in the future, I don't see big necessity for it and would say we can leave it out of MVP.

> Is it possible to erase time range after having it set?

Dates are always set to something. What would be the use case for erasing time range?

**Comment by 0x656b694d** [comment]
> > Is it possible to erase time range after having it set?
> 
> Dates are always set to something. What would be the use case for erasing time range?

To query everything collected. One would need to pick some random dates in the very past and in the future to include the whole dataset.

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/ShowProductUsage.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                variant={ButtonVariant.secondary}
+                onClick={() => {
+                    setIsModalOpen(true);
+                }}
+            >
+                Show product usage
+            </Button>
+            <Modal
+                title="Product usage"
+                description="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin blandit augue est. Duis fermentum lacinia consequat. Donec id felis elit. Donec nec nunc felis. Morbi quis enim scelerisque, ullamcorper velit."
```
Comment (0x656b694d) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/ShowProductUsage.tsx:26]: I personally like the text, but not sure if customers will appreciate it. What about this:
```suggestion
                description="The page shows the collected product usage data. The current usage is computed from the last metrics received from sensors, and can be accurate to about 5 minutes. The maximum usage is aggregated hourly."
```

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                </DescriptionListGroup>
+                <DescriptionListGroup>
+                    <DescriptionListTerm>End date</DescriptionListTerm>
+                    <DescriptionListDescription>
+                        <DatePicker
+                            value={yyyyMMddFormat(endDate)}
+                            onChange={(_str, _, date) => {
+                                if (date) {
+                                    // Add 23 hours 59 minutes to include end date completely in the request.
+                                    date.setTime(date.getTime() + (24 * 60 - 1) * 60 * 1000);
```
Comment (0x656b694d) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:118]:
```suggestion
                                    // Add 1 day to include end date completely in the request.
                                    date.setUTCDate(date.getUTCDate() + 1);
```

File: ui/apps/platform/src/types/productUsage.proto.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+export type TimeRange = {
+    from: string;
+    to: string;
+};
+
+export type MaxSecuredUnitsUsageResponse = {
+    maxNodes: number;
+    maxNodesAt: string;
+    maxCpuUnitsAt: string;
+    maxCpuUnits: number;
```
Comment (0x656b694d) [inline ui/apps/platform/src/types/productUsage.proto.ts:15]: If this file is not generated:
```suggestion
    maxNodes: number;
    maxNodesAt: string | null;
    maxCpuUnits: number;
    maxCpuUnitsAt: string | null;
```

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                    <DescriptionListDescription>
+                        {maxUsage && maxUsage?.maxCpuUnitsAt
+                            ? yyyyMMddFormat(new Date(maxUsage.maxCpuUnitsAt))
+                            : '-'}
+                    </DescriptionListDescription>
+                </DescriptionListGroup>
+                <DescriptionListGroup>
+                    <DescriptionListTerm>Nodes count collection date</DescriptionListTerm>
+                    <DescriptionListDescription>
+                        {maxUsage && maxUsage?.maxNodesAt
```
Comment (0x656b694d) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:144]:
```suggestion
                        {maxUsage?.maxNodesAt
```

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                    <DescriptionListDescription>{maxUsage.maxCpuUnits}</DescriptionListDescription>
+                </DescriptionListGroup>
+                <DescriptionListGroup>
+                    <DescriptionListTerm>Nodes count</DescriptionListTerm>
+                    <DescriptionListDescription>{maxUsage.maxNodes}</DescriptionListDescription>
+                </DescriptionListGroup>
+                <DescriptionListGroup>
+                    <DescriptionListTerm>CPU units collection date</DescriptionListTerm>
+                    <DescriptionListDescription>
+                        {maxUsage && maxUsage?.maxCpuUnitsAt
```
Comment (0x656b694d) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:136]:
```suggestion
                        {maxUsage?.maxCpuUnitsAt
```

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    const [endDate, setEndDate] = useState(new Date());
+    const [currentUsage, setCurrentUsage] = useState({
+        numNodes: 0,
+        numCpuUnits: 0,
+    } as SecuredUnitsUsage);
+    const [maxUsage, setMaxUsage] = useState({
+        maxNodes: 0,
+        maxNodesAt: '-',
+        maxCpuUnits: 0,
+        maxCpuUnitsAt: '-',
```
Comment (0x656b694d) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:37]:
```suggestion
        maxNodesAt: null,
        maxCpuUnits: 0,
        maxCpuUnitsAt: null,
```

**Review by parametalol — CHANGES_REQUESTED** [review]

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+function UsageStatisticsForm(): ReactElement {
+    const initialStartDate = new Date();
+    initialStartDate.setDate(initialStartDate.getDate() - 30);
+    const [startDate, setStartDate] = useState(initialStartDate);
+    const [endDate, setEndDate] = useState(new Date());
+    const [currentUsage, setCurrentUsage] = useState({
+        numNodes: 0,
+        numCpuUnits: 0,
+    } as SecuredUnitsUsage);
```
Comment (vjwilson) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:32]: better TS syntax:

```suggestion
    const [currentUsage, setCurrentUsage] = useState<SecuredUnitsUsage>({
        numNodes: 0,
        numCpuUnits: 0,
    });
```

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    const [maxUsage, setMaxUsage] = useState({
+        maxNodes: 0,
+        maxNodesAt: '-',
+        maxCpuUnits: 0,
+        maxCpuUnitsAt: '-',
+    } as MaxSecuredUnitsUsageResponse);
```
Comment (vjwilson) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:38]: better TS syntax:

```suggestion
    const [maxUsage, setMaxUsage] = useState<MaxSecuredUnitsUsageResponse>({
        maxNodes: 0,
        maxNodesAt: '-',
        maxCpuUnits: 0,
        maxCpuUnitsAt: '-',
    });
```

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+import {
+    downloadProductUsageCsv,
+    fetchCurrentProductUsage,
+    fetchMaxCurrentUsage,
+} from '../../../services/ProductUsageService';
+import { getAxiosErrorMessage } from '../../../utils/responseErrorUtils';
+
+function UsageStatisticsForm(): ReactElement {
+    const initialStartDate = new Date();
+    initialStartDate.setDate(initialStartDate.getDate() - 30);
```
Comment (vjwilson) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:26]: It may be better to abstract this out to a helper function, since you are mutating the variable.

We also use a datetime library `date-fns` that may be a safer way to get this initial value.

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            });
+    }, [startDate, endDate]);
+
+    return (
+        <>
+            {errorFetchingCurrent && (
+                <Alert isInline title={errorFetchingCurrent} variant="danger" />
+            )}
+            {errorFetchingMax && <Alert isInline title={errorFetchingMax} variant="danger" />}
+            <div className="pf-u-font-size-lg">Currently secured</div>
```
Comment (vjwilson) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:69]: Accessibility: this should be the appropriate heading level

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                </DescriptionListGroup>
+                <DescriptionListGroup>
+                    <DescriptionListTerm>Nodes count</DescriptionListTerm>
+                    <DescriptionListDescription>{currentUsage.numNodes}</DescriptionListDescription>
+                </DescriptionListGroup>
+            </DescriptionList>
+
+            <div className="pf-u-font-size-lg" style={{ padding: '20px 0px 0px 0px' }}>
+                Maximum secured
+            </div>
```
Comment (vjwilson) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:90]: Accessibility: this should be the appropriate heading level

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx (hunk 2 above)
Comment (vjwilson) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:118]: Good appropriate use of comment

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+            <Divider style={{ padding: '5px 0px 10px 0px' }} />
```
Comment (vjwilson) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:91]: We should be adding arbitrary spacing; instead, we should use the classes that PatternFly provides to keep a consistent look and feel throughout the app.

See https://v4-archive.patternfly.org/v4/utilities/spacing

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+                <DescriptionListGroup>
+                    <DescriptionListTerm>Nodes count collection date</DescriptionListTerm>
+                    <DescriptionListDescription>
+                        {maxUsage && maxUsage?.maxNodesAt
+                            ? yyyyMMddFormat(new Date(maxUsage.maxNodesAt))
+                            : '-'}
+                    </DescriptionListDescription>
+                </DescriptionListGroup>
+            </DescriptionList>
+            <Grid hasGutter style={{ padding: '20px 0px 0px 0px' }}>
```
Comment (vjwilson) [inline ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:150]: We should be adding arbitrary spacing; instead, we should use the classes that PatternFly provides to keep a consistent look and feel throughout the app.

See https://v4-archive.patternfly.org/v4/utilities/spacing

**Review by vjwilson — CHANGES_REQUESTED** [review]
My review is for the UI files only.

There are several places that need to be tightened up, and especially modified/checked for accessibility.

Also, we either need to add ui-e2e tests, or at least create a follow-up Jira ticket to do that, so that it doesn't stay untested.

cc: @pedrottimark

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/ShowProductUsage.tsx [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+                isOpen={isModalOpen}
+                variant={ModalVariant.large}
+                onClose={onCloseModalHandler}
+                aria-label="Product usage"
+                hasNoBodyWrapper
```
[the remark this answers is not in the document]
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/ShowProductUsage.tsx:31]: First one is with `hasNoBodyWrapper`, second is without. Doc says:
```
Flag indicating if modal content should be placed in a modal box body wrapper.
```

I personally prefer first, otherwise we get a lot of empty space around the elements.

Screenshot 2023-08-21 at 18 50 17
Screenshot 2023-08-21 at 18 49 53

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx (hunk 6 above)
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:37]: I don't think `null` is appropriate value to display in the UI, especially considering it would be rendered to empty string

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx (hunk 11 above)
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:90]: @vjwilson could you reference the accessibility guideline that determines heading levels? I couldn't find one, unfortunately

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx (hunk 2 above)
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:118]: If we do that, date will change in the form as well

File: ui/apps/platform/src/services/ProductUsageService.ts [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+            return response.data;
+        });
+}
+
+export function downloadProductUsageCsv({ from, to }: TimeRange): Promise<void> {
+    return saveFile({
+        method: 'post',
+        url: '/api/product/usage/csv',
+        data: { from, to },
+    });
```
[the remark this answers is not in the document]
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/services/ProductUsageService.ts:37]: Nice catch! At some point endpoint got changed in backend but stayed the same in the UI code

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx (hunk 11 above)
Comment (vjwilson) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx:90]: > @vjwilson could you reference the accessibility guideline that determines heading levels? I couldn't find one, unfortunately

PatternFly's docs have a good presentation, starting on this page,
https://v4-archive.patternfly.org/v4/accessibility/accessibility-fundamentals

But the particular point about headings is under the Headings guideline of this interior page,
https://v4-archive.patternfly.org/v4/accessibility/product-development-guide#what-products-should-address

> Heading text should be descriptive. **Correct heading levels should be used to communicate the outline and structure of the page. A common practice is to use a single h1 for the primary headline or logo on a page, h2s to designate major sections, and h3's in supporting subsections**
(I added the emphasis)

Basically boils down to 2 heuristics
- headings should be marked up with h1 | h2 | h3 etc. (divs and spans and visual styling don't count)
- don't skip heading levels (example, don't have h2 then h4, with no h1 or h3 anywhere, or worse h3 then h1 nested beneath it)

For this modal, it may be appropriate for the modal title to be h2, since it is its own "section" on this page, that just happens to pop up when triggered.
Then individual sections within the modal would have heading starting at h3, and sub-sections within each h3 section with have h4, and so on.

cc: @pedrottimark

File: ui/apps/platform/src/types/productUsage.proto.ts (hunk 3 above)
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/types/productUsage.proto.ts:15]: Considering `null` is automatically transformed into empty string when response is handled, I don't see why we should support additional empty value here

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx (hunk 11 above)
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx]: Thanks! Should I keep `className="pf-u-font-size-lg"` if I use `h2`/`h1`?

So far I used `h2` and without `className="pf-u-font-size-lg"` it looks pretty small(see "Currently secured" down below)

Screenshot 2023-08-21 at 19 46 30

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx (hunk 2 above)
Comment (0x656b694d) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx]: Oh, then, we should only do that when querying the API.

File: ui/apps/platform/src/types/productUsage.proto.ts (hunk 3 above)
Comment (0x656b694d) [reply ui/apps/platform/src/types/productUsage.proto.ts:15]: To make typescript happy.

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx (hunk 6 above)
Comment (0x656b694d) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx]: I thought I saw and even commented on the special logic for the display, which shows '-' if no time value is available.

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx (hunk 11 above)
Comment (vjwilson) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx]: There are still style rules from old StackRox style that may cause quirky behavior of fonts sizes (CSS is global scoped.)

You can use the PatternFly <Title> Component, 
https://v4-archive.patternfly.org/v4/components/title,
to give you an auto-complete friendly way of specifying the h-level and the size for each heading.

Comment (ivan-degtiarenko) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx]: That worked, thanks!

File: ui/apps/platform/src/types/productUsage.proto.ts (hunk 3 above)
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/types/productUsage.proto.ts:15]: There's no warning or anything to identify that typescript is "not happy"

File: ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+    }, [startDate, endDate]);
+
+    return (
+        <>
+            {errorFetchingCurrent && (
+                <Alert isInline title={errorFetchingCurrent} variant="danger" />
+            )}
+            {errorFetchingMax && <Alert isInline title={errorFetchingMax} variant="danger" />}
+            <div className="pf-u-font-size-lg">Currently secured</div>
+            <Divider style={{ padding: '5px 0px 10px 0px' }} />
```
[the remark this answers is not in the document]
Comment (ivan-degtiarenko) [reply ui/apps/platform/src/Containers/SystemHealth/UsageStatistics/UsageStatisticsForm.tsx]: Removed

**Comment by ivan-degtiarenko** [comment]
@0x656b694d Addressed all code-related feedback - let's leave all the extra buttons outside of MVP and evaluate whether we'll add them in 4.3

**Review by parametalol — APPROVED** [review]

**Review by vjwilson — APPROVED** [review]
UI is good now
